public/js/components/SourceTabs.js

```diff
@@ -30,45 +32,158 @@ function getFilename(url) {
+    selectSource: PropTypes.func.isRequired,
+    closeTab: PropTypes.func.isRequired
+  },
 
-  return (
-    dom.div({ className: "source-tabs" },
-      isEnabled("features.tabs")
-        ? sourceTabs.map(renderTab)
-        : renderTab(selectedSource)
-    )
-  );
-}
+  displayName: "SourceTabs",
+
+  getInitialState() {
+    return {
+      dropdownShown: false,
+      hiddenSourceTabs: null
+    };
+  },
+
+  componentDidUpdate() {
+    this.updateHiddenSourceTabs(this.props.sourceTabs);
+  },
+
+  /*
+   * Updates the hiddenSourceTabs state, by
+   * finding the source tabs who have wrapped and are not on the top row.
+   */
+  updateHiddenSourceTabs(sourceTabs) {
+    const sourceTabEls = this.refs.sourceTabs.children;
+    const hiddenSourceTabs = getHiddenTabs(sourceTabs, sourceTabEls);
+
+    if (!hiddenSourceTabs.equals(this.state.hiddenSourceTabs)) {
+      this.setState({ hiddenSourceTabs });
+    }
+  },
+
+  toggleSourcesDropdown(e) {
+    this.setState({
+      dropdownShown: !this.state.dropdownShown,
+    });
+  },
+
+  renderSourcesDropdown() {
+    if (!this.state.hiddenSourceTabs) {
+      return dom.div({});
+    }
+
+    return dom.div({
+      className: "sources-dropdown dropdown",
+      ref: "sourcesDropdown",
+      style: { display: (this.state.dropdownShown ? "block" : "none") }
+    },
+      dom.ul({}, this.state.hiddenSourceTabs.map(this.renderDropdownSource))
+    );
+  },
+
+  renderDropdownSource(source) {
+    const { selectSource } = this.props;
+    const url = source && source.get("url");
+    const filename = getFilename(url);
+
+    return dom.li({
+      key: source.get("id"),
+      onClick: () => {
+        selectSource(source.get("id"), { position: 0 });
+        this.toggleSourcesDropdown();
+      }
+    }, filename);
+  },
+
+  renderSourcesDropdownButon() {
+    const hiddenSourceTabs = this.state.hiddenSourceTabs;
+    if (!hiddenSourceTabs || hiddenSourceTabs.size == 0) {
+      return dom.div({});
+    }
+
+    return dom.span(
+      {
+        className: "subsettings",
+        onClick: this.toggleSourcesDropdown
+      },
+      dom.img({ src: "images/subSettings.svg" })
+    );
+  },
+
+  renderTabs() {
+    const sourceTabs = this.props.sourceTabs;
+    return dom.div(
+      { className: "source-tabs", ref: "sourceTabs" },
+      sourceTabs.map(this.renderTab)
+    );
+  },
+
+  renderTab(source) {
+    const { selectedSource, selectSource, closeTab } = this.props;
+    const url = source && source.get("url");
+    const filename = getFilename(url);
+    const active = source.equals(selectedSource);
+
+    function onClickClose(ev) {
+      ev.stopPropagation();
+      closeTab(source.get("id"));
+    }
+
+    return dom.div(
+      {
+        className: classnames("source-tab", { active }),
+        key: source.get("id"),
+        onClick: () => selectSource(source.get("id"))
+      },
+      dom.div({ className: "filename" }, filename),
+      dom.div({ onClick: onClickClose },
```

Comment:
nit: there's a mixture of styles here with the arguments. I thought we decided to bump down the second argument. But I could be convinced that this is too useful, as I've seen it elsewhere. Not a big deal for now, we can ignore this.